src/swarm/neo/util/Batch.d

```diff
@@ -0,0 +1,619 @@
+
+        Extraction iterator. Extracts records from the batch one at a time and
+        passes them to the iteration delegate.
+
+        Params:
+            dg = opApply iteration delegate
+
+        Returns:
+            0 if the iteration finished or non-0 if it was broken by the called
+
+    ***************************************************************************/
+
+    public int opApply ( int delegate ( ref Record ) dg )
+    {
+        Record record;
+        while ( this.remaining.length )
+        {
+            foreach ( ref field; record )
+                this.extractField(field);
+
+            if ( auto r = dg(record) )
+                return r;
+        }
+
+        return 0;
+    }
+
+    /***************************************************************************
+
+        Extracts a single field from the batch.
+
+        Params:
+            Field = type of field
+            field = field to extract from batch (passed by reference; value is
+                extracted directly into this argument)
+
+    ***************************************************************************/
+
+    private void extractField ( Field ) ( ref Field field )
+    {
+        static if ( isDynamicArrayType!(Field) )
+        {
+            auto len = this.extractBytes(size_t.sizeof);
+            field = cast(Field)this.extractBytes(*(cast(size_t*)len.ptr));
+        }
+        else
+        {
+            auto slice = (cast(void*)&field)[0..Field.sizeof];
+            slice[] = this.extractBytes(Field.sizeof)[];
+        }
+    }
+
+    /***************************************************************************
+
+        Consumes the specified number of bytes from the batch.
+
+        Params:
+            bytes = number of bytes to consume
+
+        Returns:
+            a slice of the consumed bytes
+
+    ***************************************************************************/
+
+    private Const!(void)[] extractBytes ( size_t bytes )
+    {
+        enforce(this.remaining.length >= bytes);
+
+        auto ret = this.remaining[0 .. bytes];
+        this.remaining = this.remaining[bytes .. $];
+        return ret;
+    }
+}
+
+/*******************************************************************************
+
+    Helper function to statically validate that a record tuple is of supported
+    types:
+        * Only 1d arrays are supported.
+        * Array elements must not have indirections.
+        * Non-array fields may not contain indirections.
+
+    Params:
+        Record = tuple of types of which a record consists
+
+    Returns:
+        true if all fields of Record are supported
+
+*******************************************************************************/
+
+private bool recordFieldsSupported ( Record ... ) ( )
+{
+    foreach ( Field; Record )
+    {
+        static if ( isDynamicArrayType!(Field) )
+        {
+            static if ( rankOfArray!(Field) > 1 )
```

Comment:
This should be implied by the `hasIndirections` check.

Reply:
For the sake of an error message that is more comprehensive than "something is wrong" I suggest using `static assert`s on the error conditions with the unsupported type and maybe even the field name in the message, and naming the function void `checkRecordFieldsSupported` or something similar.

Reply:
:+1: updated.